mitmproxy/proxy/layers/dns.py

```diff
@@ -91,24 +148,32 @@ def state_query(self, event: events.Event) -> layer.CommandGenerator[None]:
         from_client = event.connection is self.context.client
 
         if isinstance(event, events.DataReceived):
+            msgs: List[dns.Message] = []
             try:
-                msg = dns.Message.unpack(event.data)
+                if self.context.client.transport_protocol == "udp":
+                    assert self.context.server.transport_protocol == "udp"
+                    msgs.append(dns.Message.unpack(event.data))
+                elif self.context.client.transport_protocol == "tcp":
+                    assert self.context.server.transport_protocol == "tcp"
+                    self.buf.extend(event.data)
```

Comment:
@mhils can we expect to have race conditions corrupting `DNSLayer.buf` here?
for eg: `DNSLayer.buf` might have incomplete data from a previously pipelined request that might be corrupted when we receive a response from the server before we receive the rest of the query.
One solution I can think of is to have a `DNSLayer.req_buf` and `DNSLayer.resp_buf`

Reply:
@errorxyz: yes, absolutely. I didn't realize we're sharing a buffer here, we definitely need to split those. 😅

Reply:
alright👍